src/util/KinBodyParser.cpp

```diff
@@ -0,0 +1,532 @@
+
+  // Transform all the element names to lower cases because the KinBody spec is
+  // case-insensitive. We use lower cases to read the XML elements.
+  transformElementNamesToLowerCases(kinBodyDoc);
+
+  auto kinBodyEle = kinBodyDoc.FirstChildElement("kinbody");
+  if (!kinBodyEle)
+  {
+    dterr << "[KinBodyParser] KinBody document does not contain <KinBody> as "
+          << "the root element.\n";
+    return nullptr;
+  }
+
+  return readKinBody(kinBodyEle, baseUri, retriever);
+}
+
+//==============================================================================
+dart::dynamics::SkeletonPtr readKinBody(
+    tinyxml2::XMLElement* kinBodyEle,
+    const dart::common::Uri& baseUri,
+    const dart::common::ResourceRetrieverPtr& retriever)
+{
+  assert(kinBodyEle != nullptr);
+  // kinBodyEle shouldn't be nullptr at all, which always should be
+  // gauranteed because this function is only used internally to be so. If we
+  // find the nullptr case, then we should fix the caller.
+
+  auto skeleton = dart::dynamics::Skeleton::create();
+
+  // Read name attribute.
+  //
+  // If name attribute doesn't exist, getAttributeString() returns empty string.
+  // Empty name for skeleton is allowed.
+  auto name = dart::utils::getAttributeString(kinBodyEle, "name");
+  if (name.empty())
+  {
+    dtwarn << "[KinBodyParser] The <KinBody> in '" << baseUri.toString()
+           << "' doesn't have name attribute or has empty name.\n";
+  }
+  skeleton->setName(name);
+
+  // Get Body node
+  auto bodyEle = kinBodyEle->FirstChildElement("body");
+  if (bodyEle == nullptr)
+  {
+    dterr << "[KinBodyParser] KinBody document '" << baseUri.toString()
+          << "' does not contain <Body> element "
+          << "under <KinBody> element.\n";
+    return nullptr;
+  }
+
+  auto bodyNodeInfo = readBodyNodeInfo(bodyEle, baseUri);
+
+  dart::dynamics::FreeJoint::Properties jointProps;
+  jointProps.mName = DEFAULT_KINBODY_ROOT_JOINTNAME;
+  jointProps.mT_ParentBodyToJoint = bodyNodeInfo.initTransform;
+
+  auto jointAndBodyNode = skeleton->createJointAndBodyNodePair<
+      dart::dynamics::FreeJoint>(nullptr, jointProps, bodyNodeInfo.properties);
+  assert(jointAndBodyNode.first && jointAndBodyNode.second);
+
+  readGeoms(jointAndBodyNode.second, bodyEle, baseUri, retriever);
+  skeleton->resetPositions();
+  skeleton->resetVelocities();
+
+  return skeleton;
+}
+
+//==============================================================================
+BodyNodeInfo readBodyNodeInfo(
+    tinyxml2::XMLElement* bodyNodeElement,
+    const dart::common::Uri& baseUri)
+{
+  assert(bodyNodeElement != nullptr);
+  // bodyNodeElement shouldn't be nullptr at all, which always should be
+  // gauranteed because this function is only used internally to be so. If we
+  // find the nullptr case, then we should fix the caller.
+
+  BodyNodeInfo bodyNodeInfo;
+
+  dart::dynamics::BodyNode::Properties properties;
+
+  // Name attribute
+  properties.mName = "NoName";
```

Comment:
Nit: Replace this with a `DEFAULT_KINBODY_BODYNODE_NAME` constant, like you did for `KinBody` above.